src/lib.rs

```diff
@@ -2467,6 +2450,54 @@ impl Build {
             println!("{}", s);
         }
     }
+
+    #[cfg(target_os = "macos")]
+    fn fix_env_for_apple_os(&self, cmd: &mut Command) -> Result<(), Error> {
+        let target = self.get_target()?;
+        let host = self.get_host()?;
+        if host.contains("apple-darwin") && target.contains("apple-darwin") {
+            // Replace the `SDKROOT` environment variable if it's clearly set for the wrong platform, which
+            // may occur when we're linking a custom build script while targeting iOS for example.
+            // Removing is not enough, because of cc from XCode can not find include files in such case
+            if let Ok(sdkroot) = env::var("SDKROOT") {
+                if sdkroot.contains("iPhone") {
+                    let macos_sdk = self.apple_sdk_root("macosx")?;
+                    cmd.env("SDKROOT", macos_sdk);
+                }
+            }
+            // Additionally, `IPHONEOS_DEPLOYMENT_TARGET` must not be set when using the Xcode linker at
+            // "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld",
+            // although this is apparently ignored when using the linker at "/usr/bin/ld".
+            cmd.env_remove("IPHONEOS_DEPLOYMENT_TARGET");
+        }
+        Ok(())
+    }
+    #[cfg(not(target_os = "macos"))]
+    fn fix_env_for_apple_os(&self, _cmd: &mut Command) -> Result<(), Error> {
+        Ok(())
+    }
+
+    fn apple_sdk_root(&self, sdk: &str) -> Result<OsString, Error> {
+        let sdk_path = self
+            .cmd("xcrun")
```

Comment:
I think this will end up runnign `xcrun` once-per-object, which probably slows things down by a lot?

Reply:
> I think this will end up runnign `xcrun` once-per-object, which probably slows things down by a lot?

It is the same for ios_flags functions, that called every time via:
compile_object->try_get_compiler->add_default_flags,
should `Build::ios_flags` also be fixed?

Reply:
I added cache for both, target=macos and target=iphone.